### ISSUE TITLE
How to influence the join path in a query?

### ISSUE THREAD

**Issue by bentesha** [cube-js/cube.js#438]
Given I have three cubes A, B, C, and D, which are related as follows:

`A -> B -> D`
`A -> C -> D`

In the query below, how does CubeJs determine the join path to D from A? Is there a way to hint that it goes through B and not C? Is there a way to hint this in a query without having to hack schema files?

``` javascript
{
  measures: ['D.total'],
  dimensions: ['A.id', 'A.name']
}
```

It seems that it is up to CubeJs to determine how to join cubes which could impact results significantly. Anything I am missing in the docs?

**Comment by paveltiunov** [comment]
@bentesha Hey Benedict! It's a really great question! So most of the times for the case you described separate cube should be introduced because cube semantics inferred from the join. There's a really small percentage of cases where cube should be only one and join route should be selected. Currently there's no way to select such route. It's selected depending on order of measures and then dimensions and filters. We're thinking about providing such an API.

Could you please elaborate on your use case in detail? Domain description would help a lot.

**Comment by bentesha** [comment]
A quick solution is usually to introduce a cube that joins tables in the order that you prefer. But this approach is hacky and could be overkill if you are working with a lot of tables.

I think it is very important for the join path/direction to be predictable, and users should be able to dictate how this is done from the query aspect.

One way to do this is to introduce hints in queries as follows:

```javascript
{
  dimensions: ['Customer.id', 'Customer.businessName'],
  measures: ['SalesOrder.total'],
  filters: [{
    dimension: 'Product.id',
    operator: 'equals',
    values: ['abcde']
  }],
  hints: {
    joinOrder: ['Customer', 'SalesOrder', 'SalesOrderItem', 'Product']
  }
}
```
This should be backward compatible as long as `hints` is optional.

**Comment by paveltiunov** [comment]
@bentesha Could you please share what are possible joins for example you provided?

**Comment by bentesha** [comment]
The problem here is, I have created a frontend report generation tool that allows a user to pick the columns (dimensions) and measure fields they need. What I have noticed is that the order of the join would change depending (not sure on this) on the order in which the columns were selected, even though the order of the columns shouldn't actually have an impact on the way the tables are joined.

Now, there is no way to make sure the join order remains the same regardless of the order in which the columns are specified in the query.

I can write a piece of code to make sure columns are organized in a particular order in a query to get the desired effect, and then re-organizes back the columns in the order the user prefers just to make sure the join order remains the same.

But then, this doesn't seem to be necessary. There should be a way for users to influence the order in which cubes are joined, regardless of the complexity of cube relationships. Leaving everything to the engine to decide how cubes are joined might be counter-productive at times.

The solution of introducing a new cube only works when the query is known during development.
It doesn't work for dynamic queries.

**Comment by paveltiunov** [comment]
@bentesha Makes sense. And what type of UI you want to build in order to provide your users with ability to select join order?

**Comment by ifokeev** [comment]
@paveltiunov seems like 1 input and the radio button: input for the column name and the radio for INNER/LEFT.

This feature is legit, I think. Suppose, CubeJS needs to give users more custom settings.

**Comment by bentesha** [comment]
> @bentesha Makes sense. And what type of UI you want to build in order to provide your users with ability to select join order?

@paveltiunov The app our company built for a client is relying on CubeJs for report generation. It has a built-in UI for report customization allowing users to choose columns (dimensions/measures) they want and specify filters to zoom on data.

However, the behavior of CubeJs is not predictable when it comes to joining cubes, as the join order depends on the order in which dimensions are specified. So report data will vary depending on the order in which columns are displayed.

To make things even more complicated, as the application grows, newly introduced cubes might break existing queries as new cubes might introduce alternative join paths to the engine and there is no way to tell which join path CubeJs will choose.

I think the lack of this kind of customization makes CubeJs a solution that is more appealing to small-sized projects with a handful of cubes and simple relationships.

**Comment by paveltiunov** [comment]
@bentesha Totally makes sense. 

There're 2 approaches to solve this problem in general:
1. Don't allow cycles in joins. 95% of all analytics solutions go this way so the problem doesn't exist. As a workaround multiple cubes should be defined for polymorphic relation. Here's more on it: https://cube.dev/docs/polymorphic-cubes#top
2. Allow cycles and provide a way to define join for user.

We went with 2 option in mind so we're open to support this option in API. I'm just trying to figure out how do envision UI for a user so we can provide appropriate API for your use case? Are your users experienced enough to write cube joins in a text field or you want to provide them with all possible options so they can choose one?

**Comment by bentesha** [comment]
@paveltiunov 

But the problem I stated above has nothing do to with cycles. It is just that, when CubeJs sees more than one possible way to join tables, there is no way to tell which one CubeJs should pick from the query point of view.

**Comment by paveltiunov** [comment]
@bentesha Yeah. Sorry. I should be more specific about terms. By cycles I mean undirected cycles. It's basically polytree VS directed graph.

So I totally got the problem you have. Just trying to figure what's the best solution here. And it's why I'm curious how experienced your users are and what UI here will be acceptable for them?

**Comment by bentesha** [comment]
@paveltiunov My users have very little technical knowledge. They don't even know what MySQL or Postgre is. They are, however, familiar with productivity tools like MS Excel. To them, a table is just rows and columns and nothing more.

We do have a UI for choosing the type of report a user wants. For instance, "Revenue report". Essentially, each report is a cube. Users can then decide how they would like to view 'revenue', either on a per-day basis, per sales agent, per customer, per region etc. This gives them the flexibility to decide how revenue data should be presented.

Each report has a set of possible 'views'. For example, a user could choose to view revenue by a sales agent, region, customer, as times series data etc. Each view has a set of possible columns a user can choose from.

What the UI tool basically does is guiding the user in choosing a report, configure columns and filters - and the reporting tool would use this information to construct a query and send it to CubeJs

Now, the biggest problem with CubeJs is its query syntax. It is very limited. It needs to be expanded to make it possible for users to specify additional information, like hinting on how cubes should be joined etc.

Another area I think the query syntax needs improvement is specifying how filters are applied - either before or after cubes are joined.

For instance, given the cubes below

```javascript
cube('Customer', {
  sql: `select id, name from customers`,

  joins: {
    Order: {
      relationship: 'hasMany',
      sql: `${CUBE}.id = ${Order}.customerId`
    }
  },

  measures: {
    count: {
      sql: `id`,
      type: 'count'
    }
  },

  dimensions: {
    id: {
      sql: `id`,
      type: `string`,
      primaryKey: true
    },

    name: {
      sql: `name`,
      type: 'string'
    }
  }
})
```

```javascript
cube(`Order`, {
  sql: `select customerId, date, total from orders`,

  measures: {
    count: {
      type: `count`
    },

    total: {
      sql: `total`,
      type: `sum`
    }
  },

  dimensions: {
    id: {
      sql: `id`,
      type: 'number',
      primaryKey: true
    },

    customerId: {
      sql: `customerId`,
      type: `string`
    },

    date: {
      sql: `date`,
      type: 'time'
    }
  }
});
```

How do we write a query to retrieve customers who ordered less than 200 last month?

The query below does not return customers with zero orders. This is because the date filter is applied after tables are joined and would eliminate all rows with a null date value (i.e. customer with no matching orders) even though such data should be included

```javascript
{
  "dimensions": ["Customer.id", "Customer.name"],
  "measures": ["Order.total"],
  "timeDimensions": [{
    "dimension": "Order.date",
    "dateRange": "last month"
  }],
  "filters": [{
    "dimension": "Order.total",
    "operator": "lt",
    "values": ["200"]
  }]
}
```

**Comment by paveltiunov** [comment]
@bentesha Makes sense. I think we can stick with this format proposed by you for now:
```javascript
{
  dimensions: ['Customer.id', 'Customer.businessName'],
  measures: ['SalesOrder.total'],
  filters: [{
    dimension: 'Product.id',
    operator: 'equals',
    values: ['abcde']
  }],
  hints: {
    joinOrder: ['Customer', 'SalesOrder', 'SalesOrderItem', 'Product']
  }
}
```
So the implementation can add cubes to join graph from first to last one. It means there's no requirement to have direct connection between two cubes, only that there's a join path between two.

Regarding `LEFT` vs `INNER` join I believe we should stick with this one: https://github.com/cube-js/cube.js/issues/494
For cases where do you need to save `NULL` rows please see: https://github.com/cube-js/cube.js/issues/482

**Comment by maxgrv** [comment]
Hi @paveltiunov and @bentesha,
I think that we should also consider the case that the join path could be different between dimensions, measures and filters. E.g. an order cube could have a shipping address which has a city and a customer, which could also have a city. One could want to filter by shipping address city and use customer city as a dimension. 

What about specifying the join path directly in the measure/dimension/filter paths? This approach is flexible and could be optional so to be backwards compatible too.

```
{
  dimensions: ['Customer.id', 'Customer.businessName'],
  measures: ['SalesOrder.total'],
  filters: [{
    dimension: 'Costumer.SalesOrder.SalesOrderItem.Product.id',
    operator: 'equals',
    values: ['abcde']
  }],
 }
```

**Comment by bentesha** [comment]
@maxgrv I thought about this one as well. I think CubeJs should be flexible enough to allow users to either specify a full path to avoid any ambiguities, or a partial path and let CubeJs figure out the rest.

Also, are there any plans of supporting other logical operators `OR`, `NOT` etc?

**Comment by paveltiunov** [comment]
Hey @maxgrv ! It's a great point! Do you have any ideas on how to represent it in UI/metadata? 

@bentesha There's a separate issue for complex boolean logic #259.

**Comment by maxgrv** [comment]
@paveltiunov thanks! As for metadata, we use a custom metadata layer, but I think the only addition needed to standard metadata is a list of the joined cubes for each cube (could be another array property as to be backwards compatible). 

Even if not necessary, it would also be nice if the join relation could be named in order to have multiple join relation with the same cube and different FKs without recurring to inheritance. In that case, in the join path one could specify the join name instead of the cube name if there are multiple joins to the same cube (also backwards compatible with the current and proposed approach). 

As for the UI, a recursive treeview would be necessary to let the user choose the right join path.

**Comment by philippefutureboy** [comment]
Hey there!

I have a use-case that is also affected by this issue.

In my case the problem is that some of my tables support only subsets of some of the dimensions' value sets.
As a result, the order of the join matters because the LEFT JOIN of such a table that supports only a subset of the dimensions value sets results in a mask of important information that may be available on other tables, including both dimension and fact tables.

On my side, Cube.js is used internally as a way to streamline the retrieval and computation of facts using an OLAP approach, and as such I'm interested in more technical solutions.
Personally I think that a default order specification file could be a very interesting addition to the solution provided above.
The specification file would have a format similar to below (and could be part of the `cube.js` options file):

```
[
  { schema: "cube_1_name" }, // object for extensibility
  { schema: "cube_2_name" },
]
```

I would not add the capacity to specify join as part of the default file as I think it could be misleading to the end users, both technical and non-technical. Finally, the hints as provided by @bentesha could be used to override the default order.

---

For the time being I'll update the logic of my tables so that all of my tables support the full domain of all dimensions, albeit with empty records, and I'll update the measures in the schema where applicable.

**Comment by philippefutureboy** [comment]
I'd like to suggest a syntax to include join path order in the schema itself. Add a property with a 'before' or 'after' value that would hint the compiler to put the join-defined cube after the current cube in the join order. Would be overridable via the join hints in the query schema. Here are examples:

base case:

```
joins: {
    calendar: {
      relationship: `belongsTo`,
      sql: `${CUBE}.date_key = ${calendar}.date_key`,
      order: `before`
    },

    other_facts_table: {
      relationship: `hasOne`,
      sql: `
        AND ${CUBE}.date_key = ${other_facts_table}.date_key
        AND ${CUBE}.dimension_key = ${other_facts_table}.dimension_key
      `,
      order: `after`
    },
},
```

with join type:

```
joins: {
    calendar: {
      relationship: `belongsTo`,
      sql: `${CUBE}.date_key = ${calendar}.date_key`,
      order: `before`
      type: `INNER`
    },

    other_facts_table: {
      relationship: `hasOne`,
      sql: `
        AND ${CUBE}.date_key = ${other_facts_table}.date_key
        AND ${CUBE}.dimension_key = ${other_facts_table}.dimension_key
      `,
      order: `after`
      type: `RIGHT`
    },
},
```

I'm not sure if specifying the type of join in the schemas can be source of hard to track bugs since there is no clear way to know exactly what will be the join order for the query that will be run. With this in mind, I think it's better to stick to the join hints in the query for that specific property.

**Comment by philippefutureboy** [comment]
Other note:

Whenever I write SQL, I have a preference for starting my join chain with dimension tables and join all of my facts table against the dimension tables so as to guard myself from unknowingly filter facts from a table lower in the chain because a facts table earlier in the chain does not have a record for said dimension.

@paveltiunov Would you think that this is a pattern that could be supported by Cube.js? For me that would be the dream case scenario.

**Comment by philippefutureboy** [comment]
And final note for today (I swear XD):

Our data engineer warned us against including Cube.js in our stack because it adds a layer of complexity to the stack, which usually is a sign of increased potential for breakage and bugs. He especially warned us against the potential of Cube.js to do _magic things_, since when something does magic things, it means you don't necessarily understand what happens under the hood, and that leads to more unknown unknowns.

This case here is a clear case of unknown unknown and unknown source of breakage for me, and while I'm a huge fan of Cube.js I can't disagree with the fact that this type of undocumented, opaque and automated behaviour is a significant source of risk.

**Comment by bentesha** [comment]
@philippefutureboy
In a random situation, it is possible to have a query that involves two joins with the same `order` values. Your proposal is even more confusing when a query involves more than two joins. You will have at least two `before` or `after` joins. Which `before` comes `before` the other `before`?

**Comment by philippefutureboy** [comment]
@bentesha Unless I'm missing something, here's how this should play out:

* The property works like `join.relationship` property where `hasMany` is on the opposite side of `belongsTo`, the join order property on the opposite end of the join should have the opposite link direction (aka table A `before` B is same as B `after` A). A join order could be specified only on one schema and the compiler could infer the other side of the relationship automatically. However it would be preferable that the join order is explicitly defined on both ends of  a join relationship.
* If no order is specified on either side of the join, then you effectively give Cube.js carte-blanche (which is most likely some DFS or BFS algorithm)

Now my graph algorithm and expertise is limited. Here's how I would see it:

Cube.js would create a graph with nodes being schemas and edges being joins. All edges would be directed according to the order relationship (before = outgoing, after = incoming, undefined = bidirectional).
Cube.js would then find the root nodes (nodes without incoming edges, ignoring bidirectional edges) and then move down from there in a BFS or DFS fashion ignoring the bidirectional edges. Once that traversal is done, Cube.js would then do a second pass only for bidirectional edges.
As for the policy used to decide which ancestor node to use for a unidirectional path, that's a good question.
I think that the safest approach would be to join with the node with the closest to a root.

I'm no expert on the subject, so I'd suggest someone with more expertise for a more solid answer.

**Comment by bentesha** [comment]
@philippefutureboy 
For simple/straightforward scenarios, CubeJs handles join well, in a manner that is easily predictable. In my opinion, your proposal only offers a small improvement to what CubeJs is already able to do. I don't see how this can work in a complex query that involves many cubes.  i.e your proposal is too simple/limited for complex scenarios.

**Comment by philippefutureboy** [comment]
@bentesha
Sorry for the delay, your message completely flew over my head 😄 

> CubeJs handles join well, in a manner that is easily predictable

I didn't know that. Care to elaborate on the rules? Is there any documentation for that? (I saw that @hassankhan linked an issue on the subject).

> In my opinion, your proposal only offers a small improvement to what CubeJs is already able to do. I don't see how this can work in a complex query that involves many cubes. i.e your proposal is too simple/limited for complex scenarios.

Once again, care to expand? What are the limitations you speak of? Doesn't my proposed algorithm solve most (if not all) use cases? I'm 100% open to improvement on my proposed solution. I just want a way to remove the magic in the join order or at least make it really really clear to anyone without having to understand Cube.js' inner workings. Bonus points if the solution allows me to start with dimensions where possible.

**Comment by hassankhan** [comment]
Hi @philippefutureboy, I wanted to chime in with my 2 cents 😅

> Our data engineer warned us against including Cube.js in our stack because it adds a layer of complexity to the stack, which usually is a sign of increased potential for breakage and bugs. He especially warned us against the potential of Cube.js to do magic things, since when something does magic things, it means you don't necessarily understand what happens under the hood, and that leads to more unknown unknowns.

One of my favourite things about using Cube.js while working with a dedicated Data team was that the Data team could reuse their reporting SQL queries (and therefore be able control the exact order of joins), and the Engineering team could use those same queries as the `sql` property in a data schema. This helped remove a lot of "magic" involved, since we were quite literally using Cube.js purely for caching/pre-aggregating and as a query layer. I totally get the purity(?) of being able to do everything entirely through Cube.js (trust me 😂), but doing the above helped us keep everyone happy AND roll out Cube.js in production.

With regards to a possible solution, the most flexible solution seems to be adding a property to the query format, as per @bentesha and @maxgrv's suggestions. I'd like to further suggest that we could use `->` and `=>` as join indicators, `->` indicating a `LEFT JOIN` and `=>` indicating an `INNER JOIN`. Some example scenarios are below.

**All `LEFT JOIN`s:**

```json
{
  "dimensions": ["Customer.id", "Customer.businessName"],
  "measures": ["SalesOrder.total"],
  "filters": [{
    "dimension": "Customer->SalesOrder->SalesOrderItem->Product.id",
    "operator": "equals",
    "values": ["abcde"]
  }],
}
```

`INNER JOIN` `SalesOrder` and `SalesOrderItem`:

```json
{
  "dimensions": ["Customer.id", "Customer.businessName"],
  "measures": ["SalesOrder.total"],
  "filters": [{
    "dimension": "Customer=>SalesOrder=>SalesOrderItem->Product.id",
    "operator": "equals",
    "values": ["abcde"]
  }],
}
```

`INNER JOIN` `Customer`:

```json
{
  "dimensions": ["Customer.id", "Customer.businessName"],
  "measures": ["SalesOrder.total"],
  "filters": [{
    "dimension": "=>Customer->SalesOrder->SalesOrderItem->Product.id",
    "operator": "equals",
    "values": ["abcde"]
  }],
}
```

Would love to hear thoughts/feedback 😄

**Comment by MattGson** [comment]
@philippefutureboy 

> Once again, care to expand? What are the limitations you speak of? Doesn't my proposed algorithm solve most (if not all) use cases?

As far as I can tell, your solution has a major limitation which is that you cannot chose the join ordering at query time. The other limitation I see is that you have a decentralised specification of join order. This means it is difficult to predict what the output query will be without looking through all the cubes involved.

**Comment by philippefutureboy** [comment]
Happy New Year to all of ya! ✨  🌟 

@hassankhan 

> Hi @philippefutureboy, I wanted to chime in with my 2 cents 😅

To be in the clear, I didn't have any intention to sound abrasive/rude/angry, so sorry if it may have sounded like that! 😬  😅 

> One of my favourite things about using Cube.js while working with a dedicated Data team was that the Data team could reuse their reporting SQL queries (and therefore be able control the exact order of joins), and the Engineering team could use those same queries as the sql property in a data schema. This helped remove a lot of "magic" involved, since we were quite literally using Cube.js purely for caching/pre-aggregating and as a query layer. [...]

I see! So basically you added an extra view before Cube.js to join all the information you wanted and force the join order and you still kept the relations with dimension tables?
I came to a similar conclusion on my side. I will either:

* create an extra step for final views to be used by Cube.js that allow me to make each view have at least one record for each relevant dimension combination or
* dynamically compose the SQL that serves as the base of the schemas using a COMPILE_CONTEXT with the cubejs query exposed.

RE: Alternative solution for the INNER JOIN / LEFT JOIN hints

I prefer @bentesha's solution. Wouldn't your solution allow someone to specify different join paths for each dimension/measure/filter? If so, how would the engine know which path to pick?

@MattGSon 

> As far as I can tell, your solution has a major limitation which is that you cannot chose the join ordering at query time. The other limitation I see is that you have a decentralised specification of join order. This means it is difficult to predict what the output query will be without looking through all the cubes involved.

a) you cannot chose the join ordering at query time:

I don't think my implementation is incompatible if you allow join hint at query time to override the configuration of the schema. This being said, good point cause even with support for query-time override, I think that would increase the complexity and opaqueness of Cube.js' queries way too much. In that case, it may be better to stick with a query-time join hint.

b) decentralised specification of join order

Definitely an issue, and makes it hard to predict. My solution to that would be to add a way to interactively test join paths in development, either via the dashboard or via an endpoint (`/sql` or otherwise). Once again as for (a), this increases complexity and goes against the whole point of the suggestion - clarity and simplicity.

I must agree with you, query-time join hints only seems to be the way to go.

**Comment by philippefutureboy** [comment]
Hey all!

Following up on attempting to implement a solution similar to what @hassankhan proposed:

In order to circumvent the fact that some records are filtered due to the LEFT JOIN + JOIN order, I've used the following:

```js
cube(`charge_facts`, {
  sql: `
    SELECT
        schedule_track_calendar_view.tenant_key,
        schedule_track_calendar_view.date_key,
        schedule_track_calendar_view.schedule_track_key,
        receivable_charge_facts_keyview.* EXCEPT(tenant_key, date_key, schedule_track_key)
    FROM
        \`${dataset_key}\`.schedule_track_calendar_view
        LEFT JOIN
         \`${dataset_key}\`.receivable_charge_facts_keyview
            ON schedule_track_calendar_view.clinic_key = receivable_charge_facts_keyview.clinic_key
            AND schedule_track_calendar_view.date_key = receivable_charge_facts_keyview.date_key
            AND schedule_track_calendar_view.schedule_track_key = receivable_charge_facts_keyview.schedule_track_key
  `,
  // ...
})

which compiles to:

```sql
SELECT
      `schedule_track_calendar`.clinic_key `schedule_track_calendar__clinic_key`, CAST(`schedule_track_calendar`.date_key AS TIMESTAMP) `schedule_track_calendar__date_key`, `schedule_track`.label `schedule_track__label`, sum(`charge_facts`.amount_cad) `charge_facts__charges_amount`
    FROM
      (
    SELECT
        schedule_track_calendar_view.clinic_key,
        schedule_track_calendar_view.date_key,
        schedule_track_calendar_view.schedule_track_key,
        receivable_charge_facts_keyview.* EXCEPT(clinic_key, date_key, schedule_track_key)
    FROM
    ...
    ) AS `charge_facts`
```

Since in more complex cases cube.js seem to wrap the sql property as a subquery without modifying the `*` components, the `SELECT receivable_charge_facts_keyview.* EXCEPT(tenant_key, date_key, schedule_track_key)` allows columnar SQL engines (BigQuery in this case) to optimize which fields it will read out of the `receivable_charge_facts_keyview.*` based on the top level SELECT statement (in this case `amount_cad`). Unfortunately this feature of BQ is not documented, so YMMV.

Cheers!

**Comment by andrekiba** [comment]
Hello, any news about this? Is there a plan to support the possibility to influence the join path when more than one is possible?

**Comment by emahuni** [comment]
yeah, surely this is one big issue I am having with CubeJs, it goes the wrong way, especially if there are 2 joins that are correct for different scenarios. EG: toPhone and fromPhone are cubes that both join to a Phone Cube, but we want to join through toPhone at some point and fromPhone at some. Influencing this at query time like the author of the thread suggested is exactly how this can be simplified.

**Comment by rpaik** [comment]
Closing this issue as this can now be addressed with [Views](https://cube.dev/blog/introducing-views).
